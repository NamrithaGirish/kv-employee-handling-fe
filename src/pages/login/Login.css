.row{
    display: flex;
    height: 100vh; 
}
.column1{
    display: flex; 
    background-color: rgba(3, 174, 238, 0.1);
    width:50%;
    justify-content: center;
    align-items: center;  
    margin: 20px;
    border-radius: 10px;
}
.login_image{
    object-fit: cover;
    border-radius: 50%;
    width:80%;
    height:80%; 
    /* padding: 10px; */
    /* margin: 10px; */
}
.username-clear-holder{
    width:100%;
    /* justify-content: center; */
    position: relative;
    display: inline-block;
}
.login-box{
    width:100%;
    /* display: inline; */
}
.clear{
    position: absolute;
    top:0;
    right: 0;
    background-color: #00aaf4;
    border-style: solid;
    border-color: #B4B7BD;
    height: 50px;
    width: 20%;
    color: white; 
    font-weight: 40;
    border-width: 1px;
    border-radius: 5px;
    
    
}
.column2 {
    background-color: white;
    width: 50%;
    height: 100%;
    display: flex;
    justify-content: center; /*alligns centre along h*/
    align-items: center;  /*alligns centre along v */
}

.column2 img{
    width:30%;
    margin-bottom: 20px;
    margin-left: 24%;
    background-color: white;
}

.form{
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    gap:15px;
    width: 50%;
    margin-left: auto;
    margin-right: auto;
}
.show-password{
    display: flex;
    flex-direction: row;
    margin-right: auto;
    margin-left: 0;
    /* justify-content: left;
    align-items:left; */
    /* width:100% */
    gap:5px;
}

