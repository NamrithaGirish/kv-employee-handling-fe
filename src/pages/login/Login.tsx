import React, { useEffect, useRef, useState } from "react";
import { Button } from "../../components/button/Button";
import { InputBox } from "../../components/inputBox/InputBox";
import "./Login.css";
import loginImage from "../../assets/kv-login.jpeg";
import kvLogo from "../../assets/kv-logo.png";
import { SetLocalStorage } from "../../hooks/LocalStorage";
import { useNavigate } from "react-router-dom";
import { useLoginMutation } from "../../api-service/auth/login.api";
export const Login = () => {
	const [login, { isLoading }] = useLoginMutation();
	// const mouseTrack = MouseTracker();
	const [username, setUsername] = useState("");
	let [password, setPassword] = useState("");
	let [usernameValue, setUsernameValue] = useState("");
	let [passwordValue, setPasswordValue] = useState("");
	let [clearButtonDisable, setclearButtonDisable] = useState(true);
	let navigate = useNavigate();

	const localStorageHook = SetLocalStorage("show-password");
	const authlocalStorageHook = SetLocalStorage("isLoggedIn");

	useEffect(() => {
		if (authlocalStorageHook.value === true) {
			navigate("/employee");
		}
	}, []);
	const handleLogin = async () => {
		console.log("username", username, "password", password);
		await login({ email: username, password: password })
			.unwrap()
			.then((response) => {
				localStorage.setItem("token", response.accessToken);
				console.log("Access token : ", response.accessToken);
				authlocalStorageHook.setter(true);
				const userId = JSON.parse(
					atob(response.accessToken.split(".")[1])
				);
				console.log("userId", userId.id);
				localStorage.setItem("UserId", userId.id);
				navigate(`/employee`);
			})
			.catch((error) => {
				console.log("Login Error : ", error);
				const errMsg = error.data.error || error.data.message;
				authlocalStorageHook.setter(false);
				setPasswordValue(errMsg);
			});
	};
	const usernameRef = useRef<HTMLInputElement>(null);
	useEffect(() => {
		// console.log(username);
		if (username.length > 0) {
			setclearButtonDisable(false);
		} else {
			setclearButtonDisable(true);
		}

		if (username.length > 20) {
			setUsernameValue("Username too lengthy");
		}
	}, [username]);
	useEffect(() => {
		if (usernameRef.current) {
			console.log("Username Reference : ", usernameRef);
			usernameRef.current.focus();
		}
		return () => {
			console.log("Returning from reference for username input");
		};
	}, [username]);
	return (
		<div className="row">
			<div className="column1">
				<img
					src={loginImage}
					alt="Login Image"
					className="login_image"
				/>
			</div>
			<div className="column2">
				<div className="login-box">
					<img src={kvLogo} alt="Keyvalue Logo" className="logo" />
					<div className="form">
						<div className="username-clear-holder">
							<InputBox
								value={username}
								text={"Username"}
								classname="username"
								onchange={(
									e: React.ChangeEvent<HTMLInputElement>
								) => setUsername(e.target.value)}
								substring={usernameValue}
								reference={usernameRef}
								endAdornment={
									<Button
										classname="clear"
										type="button"
										text="Clear"
										onClickFunction={() => {
											setUsername("");
											setclearButtonDisable(true);
										}}
										disabled={clearButtonDisable}
									/>
								}
							/>
						</div>
						{/* <label>{value}</label> */}
						{/* <p>
                               x= {mouseTrack.x} y={mouseTrack.y}
                            </p> */}
						<div className="username-clear-holder">
							<InputBox
								text="Password"
								classname="textinput"
								substring={passwordValue}
								onchange={(
									e: React.ChangeEvent<HTMLInputElement>
								) => setPassword(e.target.value)}
								inputType={
									localStorageHook.value ? "text" : "password"
								}
							/>
						</div>
						<div className="show-password">
							<InputBox
								value={password}
								text="checkbox"
								classname="password-checkbox"
								inputType="checkbox"
								onchange={(
									e: React.ChangeEvent<HTMLInputElement>
								) => localStorageHook.setter(e.target.checked)}
								checked={localStorageHook.value}
							/>
							Show password
						</div>
						<Button
							type={"submit"}
							text="Log In"
							classname="login"
							disabled={isLoading}
							onClickFunction={handleLogin}
						/>
					</div>
				</div>
			</div>
		</div>
	);
};
