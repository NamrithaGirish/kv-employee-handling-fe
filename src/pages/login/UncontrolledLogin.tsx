import React, { useEffect, useRef } from "react";
import { Button } from "../../components/button/Button";
import { InputBox } from "../../components/inputBox/InputBox";
import "./Login.css";
const print = () => {
	console.log("Hello");
};
export const UncontrolledLogin = () => {
	const usernameRef = useRef<HTMLInputElement>(null);
	const passwordRef = useRef<HTMLInputElement>(null);
	const clearButtonRef = useRef<HTMLButtonElement>(null);
	// const handleLogin = (event: React.FormEvent<HTMLFormElement>) => {
	// 	event.preventDefault();
	// 	const form = event.currentTarget;
	// 	const formdata = new FormData(form);
	// 	const username = formdata.get("username") as string;
	// 	const password = formdata.get("password") as string;
	// 	console.log(username, password);
	// };
	const handleRefLogin = (event: React.FormEvent<HTMLFormElement>) => {
		event.preventDefault();
		console.log(usernameRef.current?.value);
	};
	const clearUsername = () => {
		console.log("hello");
		if (!usernameRef.current) return;
		usernameRef.current.value = "";
		clearButtonRef.current
			? (clearButtonRef.current.disabled = true)
			: false;
	};
	const updateClearButton = (value: string) => {
		if (!clearButtonRef.current) return;
		if (value.length > 0) {
			clearButtonRef.current.disabled = false;
		}
		if (value.length <= 0) {
			clearButtonRef.current.disabled = true;
		}
		console.log(clearButtonRef.current.onclick);
		clearButtonRef.current.onclick = clearUsername;
		console.log(clearButtonRef.current.onclick);
	};
	const clearPassword = () => {
		if (!passwordRef.current) return;
		passwordRef.current.value = "";
	};
	useEffect(() => {
		if (usernameRef.current) {
			console.log(usernameRef);
			usernameRef.current.focus();
		}
		return () => {
			console.log("in ref return");
		};
	}, []);

	return (
		<div className="row">
			<div className="column1">
				<img
					src="./src/assets/kv-login.jpeg"
					alt="Login Image"
					className="login_image"
				/>
			</div>
			<div className="column2">
				<div className="login-box">
					<img
						src="./src/assets/kv-logo.png"
						alt="Keyvalue Logo"
						className="logo"
					/>
					<form className="form" onSubmit={handleRefLogin}>
						<InputBox
							text={"Username"}
							classname="username"
							onchange={(
								e: React.ChangeEvent<HTMLInputElement>
							) => updateClearButton(e.target.value)}
							reference={usernameRef}
							name="username"
							endAdornment={
								<button
									ref={clearButtonRef}
									disabled={true}
									className="clear"
									type={"button"}
									onClick={clearUsername}
								>
									Clear
								</button>
							}
						/>
						<InputBox
							text="Password"
							classname="textinput"
							reference={passwordRef}
							name="password"
							endAdornment={
								<Button
									text="clear"
									classname="clear"
									type={"button"}
									functionName={clearPassword}
								/>
							}
						/>
						<Button
							text="Log In"
							classname="login"
							functionName={print}
						/>
					</form>
				</div>
			</div>
		</div>
	);
};
