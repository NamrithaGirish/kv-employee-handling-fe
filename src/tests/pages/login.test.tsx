import { describe, it, expect, beforeEach, vi } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import userEvent from "@testing-library/user-event";
import { Login } from "../../pages/login/Login";
// Mock react-router-dom's useNavigate
const mockNavigate = vi.fn();
vi.mock("react-router-dom", async () => {
	const actual = await vi.importActual("react-router-dom");
	return {
		...actual,
		useNavigate: () => mockNavigate,
	};
});
// Mock the store
const mockStore = configureStore({
	reducer: {
		employee: (state = { employees: [] }) => state,
		department: (state = { departments: [] }) => state,
		employeeApi: (state = {}) => state,
	},
});
// Mock the login mutation
const mockLoginMutation = vi.fn();
vi.mock("../../api-service/auth/login.api", () => ({
	useLoginMutation: () => [
		() => ({
			unwrap: () => mockLoginMutation(),
		}),
		{ isLoading: false },
	],
}));
describe("Login Page", () => {
	const renderLogin = () => {
		return render(
			<Provider store={mockStore}>
				<Login />
			</Provider>
		);
	};
	beforeEach(() => {
		vi.clearAllMocks();
	});
	it("should match snapshot", () => {
		const { container } = renderLogin();
		expect(container).toMatchSnapshot();
	});
	// Test 2: Render login form with all required elements
	it("should render login form with all required elements", () => {
		renderLogin();

		// Check if username input is present
		const usernameInput = screen.getByPlaceholderText("Username");
		expect(usernameInput).toBeInTheDocument();

		// Check if password input is present
		const passwordInput = screen.getByPlaceholderText("Password");
		expect(passwordInput).toBeInTheDocument();

		// Check if submit button is present
		const submitButton = screen.getByRole("button", { name: /log in/i });
		expect(submitButton).toBeInTheDocument();
		expect(submitButton).toHaveAttribute("type", "submit");
	});
	// it("should show error when username is too long", async () => {
	// 	renderLogin();
	// 	const usernameInput = screen.getByPlaceholderText("Username");
	// 	// Type a username longer than 10 characters
	// 	await userEvent.type(usernameInput, "a".repeat(11));
	// 	// expect(screen.getByText("Username invalid")).toBeInTheDocument();
	// 	// Check if error message is displayed
	// 	// const container = getByTestId("subscript-username")
	// 	const element = screen.getAllByText("Username invalid");
	// 	expect(element).toHaveLength(2);
	// });
	it("should handle successful login", async () => {
		mockLoginMutation.mockResolvedValueOnce({ token: "fake-token" });

		renderLogin();

		// Fill in the form
		const usernameInput = screen.getByPlaceholderText("Username");
		const passwordInput = screen.getByPlaceholderText("Password");
		const submitButton = screen.getByRole("button", { name: /log in/i });

		await userEvent.type(usernameInput, "test@example.com");
		await userEvent.type(passwordInput, "password123");

		// Submit the form
		await userEvent.click(submitButton);

		// Wait for the login mutation to resolve and navigation to occur
		await waitFor(() => {
			expect(mockLoginMutation).toHaveBeenCalled();
			expect(mockNavigate).toHaveBeenCalledWith("/employee");
		});
	});
	it("should handle failure login", async () => {
		mockLoginMutation.mockRejectedValueOnce({
			data: { message: "Invalid username or password" },
		});

		renderLogin();

		// Fill in the form
		const usernameInput = screen.getByPlaceholderText("Username");
		const passwordInput = screen.getByPlaceholderText("Password");
		const submitButton = screen.getByRole("button", { name: /log in/i });

		await userEvent.type(usernameInput, "a");
		await userEvent.type(passwordInput, "a");

		// Submit the form
		await userEvent.click(submitButton);

		// Wait for the login mutation to resolve and navigation to occur
		await waitFor(() => {
			expect(mockLoginMutation).toHaveBeenCalled();
			const element = screen.getByText("Invalid username or password");
			expect(element).toBeInTheDocument();
		});
	});
	it("focus on username on mount", () => {
		renderLogin();

		const usernameInput = screen.getByPlaceholderText("Username");
		expect(document.activeElement).toBe(usernameInput);
	});
});
