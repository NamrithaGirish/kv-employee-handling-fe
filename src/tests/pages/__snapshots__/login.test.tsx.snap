// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login Page > should match snapshot 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="column1"
    >
      <img
        alt="Login Image"
        class="login_image"
        src="./src/assets/kv-login.jpeg"
      />
    </div>
    <div
      class="column2"
    >
      <div
        class="login-box"
      >
        <img
          alt="Keyvalue Logo"
          class="logo"
          src="./src/assets/kv-logo.png"
        />
        <div
          class="form"
        >
          <div
            class="username-clear-holder"
          >
            <div
              class="input-box"
            >
              <input
                class="username"
                id="Username"
                name="Username"
                placeholder="Username"
                type="text"
                value=""
              />
              <button
                class="clear"
                disabled=""
                type="button"
              >
                Clear
              </button>
            </div>
          </div>
          <div
            class="username-clear-holder"
          >
            <div
              class="input-box"
            >
              <input
                class="textinput"
                id="Password"
                name="Password"
                placeholder="Password"
                type="password"
              />
            </div>
          </div>
          <div
            class="show-password"
          >
            <div
              class="input-box"
            >
              <input
                class="password-checkbox"
                id="checkbox"
                name="checkbox"
                placeholder="checkbox"
                type="checkbox"
                value=""
              />
            </div>
            Show password
          </div>
          <button
            class="login"
            type="submit"
          >
            Log In
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
